### ISSUE TITLE
Large number of file open during unit test testMakeTrackValidationPlots in ROOT628

### ISSUE THREAD

**Issue by smuzaffar** [cms-sw/cmssw#43077]
Hi,
  I noticed that unit tests, during recent ROOT 6.28 PRs, are failing as there were too many open file descriptors [a]. Looks like, for ROOT6.28  unit test `Validation/RecoTrack/test/testMakeTrackValidationPlots` tries to open too many files e.g. running [b] on lxplus8 shows that there were over 37K file open calls in root 6.28 IBs while there were only 4K calls in normal ROOT 6.26 based IBs. ROOT630 and ROOT master based IBs also shows a lot open files issue. 

[a]  https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6a8509/35311/unitTests/failed.html
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6a8509/35311/unitTests/src/Alignment/OfflineValidation/test/DMRall/testing.log
```
./cmsRun: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: Error 24
Failure running DMR single configuration standalone: status 127
```

/var/log/messages contains entries like
```
Oct 21 16:07:35 cmsbuild114.cern.ch cvmfs2[6642]: (unpacked.cern.ch) open file descriptor limit exceeded
Oct 21 16:07:35 cmsbuild114.cern.ch cvmfs2[6642]: (unpacked.cern.ch) open file descriptor limit exceeded
Oct 21 16:07:35 cmsbuild114.cern.ch cvmfs2[6642]: (unpacked.cern.ch) open file descriptor limit exceeded
Oct 21 16:07:35 cmsbuild114.cern.ch cvmfs2[6642]: (unpacked.cern.ch) open file descriptor limit exceeded
Oct 21 16:07:35 cmsbuild114.cern.ch cvmfs2[6642]: (unpacked.cern.ch) open file descriptor limit exceeded
Oct 21 16:07:35 cmsbuild114.cern.ch cvmfs2[6642]: (unpacked.cern.ch) open file descriptor limit exceeded
Oct 21 16:07:35 cmsbuild114.cern.ch cvmfs2[6642]: (unpacked.cern.ch) open file descriptor limit exceeded
```

[b] lxplus8

- ROOT 6.26 Based IB
```
> cd /tmp/$(whoami)
> source /cvmfs/cms.cern.ch/cmsset_default.sh
> scram p CMSSW_13_3_X_2023-10-20-2300
> cd CMSSW_13_3_X_2023-10-20-2300
> cmsenv
> strace -e trace=open,openat,close,read,write,connect,accept -o run.log makeTrackValidationPlots.py ~muzaffar/public/root628/test.root
> grep '^open' run.log |wc -l
3933
```

- ROT 6.28 Based IB
```
> cd /tmp/$(whoami)
> source /cvmfs/cms.cern.ch/cmsset_default.sh
> scram p CMSSW_13_3_ROOT628_X_2023-10-20-2300
> cd CMSSW_13_3_ROOT628_X_2023-10-20-2300
> cmsenv
> strace -e trace=open,openat,close,read,write,connect,accept -o run.log makeTrackValidationPlots.py ~muzaffar/public/root628/test.root
> grep '^open' run.log |wc -l
37636
```

**Comment by smuzaffar** [comment]
Number of files open during this unit tests for ROOT 6.26 based IBs are [a] while for ROOT628 (6.20 and master) IBs are [b]

[a]
```
grep '^open' run.log | grep -v 'No such file or directory' | grep -v 'Read-only file system' | sed -e 's|.*, "||;s|".*||' | sort | grep '/cvmfs/' | cut -d/ -f9 | sort | uniq -c  | sort -n | tail -5
      4 coral
      6 gcc
    202 python3
    266 root
    367 cmssw
```

[b]
```
grep '^open' run.log | grep -v 'No such file or directory' | grep -v 'Read-only file system' | sed -e 's|.*, "||;s|".*||' | sort | grep '/cvmfs/' | cut -d/ -f9 | sort | uniq -c  | sort -n | tail -5
    530 sherpa
    541 llvm
    987 gcc
   2731 root
  16034 cmssw
```

**Comment by smuzaffar** [comment]
By the way, running `makeTrackValidationPlots.py ~muzaffar/public/root628/test.root` ( which is what run by [testMakeTrackValidationPlots](https://github.com/cms-sw/cmssw/blob/master/Validation/RecoTrack/test/test_makeTrackValidationPlots.sh#L11) ) shows that it starts over 150 processes [a]. Is this an expected behavior of this  unity test?


[a]
```
> ps -u muzaffar -o pid,cmd --forest
 876671      \_ python3 /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_13_3_ROOT628_X_2023-10-22-2300/bin/el8_amd64_gcc12/makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628/test.root
 876873      |   \_ python3 /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_13_3_ROOT628_X_2023-10-22-2300/bin/el8_amd64_gcc12/makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628/test.root
 876878      |   \_ python3 /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_13_3_ROOT628_X_2023-10-22-2300/bin/el8_amd64_gcc12/makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628/test.root
 876880      |   \_ python3 /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_13_3_ROOT628_X_2023-10-22-2300/bin/el8_amd64_gcc12/makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628/test.root
 876882      |   \_ python3 /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_13_3_ROOT628_X_2023-10-22-2300/bin/el8_amd64_gcc12/makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628/test.root
 876884      |   \_ python3 /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_13_3_ROOT628_X_2023-10-22-2300/bin/el8_amd64_gcc12/makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628/test.root
 876886      |   \_ python3 /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_13_3_ROOT628_X_2023-10-22-2300/bin/el8_amd64_gcc12/makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628/test.root
 876888      |   \_ python3 /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_13_3_ROOT628_X_2023-10-22-2300/bin/el8_amd64_gcc12/makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628/test.root
 876890      |   \_ python3 /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_13_3_ROOT628_X_2023-10-22-2300/bin/el8_amd64_gcc12/makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628/test.root
 876892      |   \_ python3 /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_13_3_ROOT628_X_2023-10-22-2300/bin/el8_amd64_gcc12/makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628/test.root
```
```
> ps -u muzaffar -o pid,cmd --forest | grep python3  |wc
    157     940   36074
```

**Comment by makortel** [comment]
Tagging @cms-sw/tracking-pog-l2 as well

**Comment by slava77** [comment]
> Is this an expected behavior of this unit test?

yes, it's expected. I can't say that it's desired or well justified.

If there is a suggestion on how to restrict `multiprocessing.Process` used in Validation/RecoTrack/python/plotting/validation.py from starting a new one, it would resolve the issue more reliably.

For my own tests I'm using `ulimit -n 4096` in the shell before running

**Comment by smuzaffar** [comment]
running `strace -f` to follow forks shows following number of `openat` calls

- root 6.26: 7K
- root 6.28: 908K
- root 6.30: 2877K 
- root master: 50K

**Comment by mmusich** [comment]
> I can't say that it's desired or well justified.

as a reminder the unit test was introduced back then in https://github.com/cms-sw/cmssw/pull/35957, because of the script getting recurrently broken by other developments. IIRC `multiprocessing` was indeed one of the reasons it stopped working once (https://github.com/cms-sw/cmssw/issues/35827).

**Comment by slava77** [comment]
> running `strace -f` to follow forks shows following number of `openat` calls
> 
>     * root 6.26: 7K
> 
>     * root 6.28: 908K
> 
>     * root 6.30: 2877K
> 
>     * root master: 50K

uhm, I'm not sure what these counts are. Is it for simultaneously opened files or just a count of attempts to open files?

For the latter, could there be some issue with the search paths? (e.g. repeated attempts at files that do not exist)

**Comment by makortel** [comment]
> looks like root 628 and above are reading/opening all the shared libraries in LD_LIBRARY_PATH

@smuzaffar Have you happened to check what happens in `cmsRun`? In general I'd think ROOT "reading/opening all the shared libraries in LD_LIBRARY_PATH" would be bad behavior in general.

**Comment by smuzaffar** [comment]
I am running a workflow

**Comment by smuzaffar** [comment]
@slava77 @mmusich , I opened https://github.com/cms-sw/cmssw/pull/43096 to control the number of parallel process for validation plots. Default is to use #CPUs jobs in parallel. This will help a bit with number of simultaneously open file descriptors but we still have to see why root 6.28 and above are opening too many files.

**Comment by smuzaffar** [comment]
For ROOT 6.26, when a [new process](https://github.com/cms-sw/cmssw/blob/master/Validation/RecoTrack/python/plotting/validation.py#L1205) is started to do the plotting then the stack for each opened shared library looks like [a] and only few (less than 100) root libs and macros were opened. But for ROOT 6.28 (and above) the stack shows something like [b] and all the shared libs from LD_LIBRARY_PATH and system directory (e.g. /usr/lib64) were opened (over 22K files were opened).

Looks like these following calls are responsible for opening all shared libs in root 6.28 and above
```
libCling.so(cling::Dyld::ContainsSymbol((anonymous namespace)::LibraryPath const*, llvm::StringRef, unsigned int) const+0x70) [0xb736e0]
libCling.so(cling::Dyld::searchLibrariesForSymbol[abi:cxx11](llvm::StringRef, bool)+0x307) [0xb76cf7]
libCling.so(cling::DynamicLibraryManager::searchLibrariesForSymbol[abi:cxx11](llvm::StringRef, bool) const+0x13) [0xb77053]
libCling.so(AutoloadLibraryGenerator::tryToGenerate(llvm::orc::LookupState&, llvm::orc::LookupKind, llvm::orc::JITDylib&, llvm::orc::JITD
```

FYI @vgvassilev  @Axel-Naumann

`strace -ff -f -k -e trace=open,openat -o run.log makeTrackValidationPlots.py /afs/cern.ch/user/m/muzaffar/public/root628//test.root` command was used to generate the following stacktraces


[a] /afs/cern.ch/user/m/muzaffar/public/root628/stack/root-6.26.11.log
```
openat(AT_FDCWD, "<inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libGpad.so", O_RDONLY) = 21
 > /usr/lib64/libc-2.28.so(open+0x52) [0x11f9e2]
 > /usr/lib64/libc-2.28.so(_IO_file_open+0x25) [0x92485]
 > /usr/lib64/libc-2.28.so(_IO_file_fopen@@GLIBC_2.2.5+0xea) [0x9263a]
 > /usr/lib64/libc-2.28.so(__fopen_internal+0x7c) [0x8679c]
 > /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02808/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/lib64/libstdc++.so.6.0.30(std::__basic_file<char>::open(char const*, std::_Ios_Openmode, int)+0x33) [0xce1a3]
 > <inst>/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/lib64/libstdc++.so.6.0.30(std::basic_filebuf<char, std::char_traits<char> >::open(char const*, std::_Ios_Openmode)+0x3d) [0x11340d]
 > <inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libCling.so(cling::DynamicLibraryManager::isSharedLibrary(llvm::StringRef, bool*) [clone .constprop.0]+0x1d2) [0x698a62]
 > <inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libCling.so(cling::DynamicLibraryManager::lookupLibrary[abi:cxx11](llvm::StringRef, llvm::SmallVector<llvm::StringRef, 2u>, llvm::SmallVector<llvm::StringRef, 2u>, llvm::StringRef, bool) const+0x77) [0x699bd7]
 > <inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libCling.so(TCling::Load(char const*, bool)+0x159) [0x5dd2c9]
 > <inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libCore.so(TSystem::Load(char const*, char const*, bool)+0x738) [0x20f348]
 > <inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libCore.so(TROOT::LoadClass(char const*, char const*, bool)+0x6b) [0x1997fb]
 > <inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libCling.so(TCling::ShallowAutoLoadImpl(char const*)+0x2d1) [0x5dad21]
 > <inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libCling.so(TCling::DeepAutoLoadImpl(char const*, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, bool)+0x280) [0x5f2740]
 > <inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libCling.so(TCling::AutoLoad(char const*, bool)+0x283) [0x5f2d03]
 > <inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libCore.so(TClass::GetClass(char const*, bool, bool, unsigned long, unsigned long)+0x4a8) [0x29e868]
 > <inst>/lcg/root/6.26.11-2d833eecc2c1ccfc13843c9f9618acfd/lib/libcppyy_backend3_9.so(Cppyy::GetScope(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xf8) [0x1
```

[b] /afs/cern.ch/user/m/muzaffar/public/root628/stack/root-6.28.09.log
```
openat(AT_FDCWD, "<inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libASImage.so", O_RDONLY|O_CLOEXEC) = 8
 > /usr/lib64/libpthread-2.28.so(__open+0x4f) [0x1221f]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::sys::fs::openFile(llvm::Twine const&, int&, llvm::sys::fs::CreationDisposition, llvm::sys::fs::FileAccess, llvm::sys::fs::OpenFlags, unsigned int)+0xbb) [0x440036b]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::sys::fs::openNativeFileForRead(llvm::Twine const&, llvm::sys::fs::OpenFlags, llvm::SmallVectorImpl<char>*)+0x35) [0x4400695]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::MemoryBuffer::getFile(llvm::Twine const&, bool, bool, bool)+0x28) [0x436f7c8]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::object::ObjectFile::createObjectFile(llvm::StringRef)+0x3d) [0x3da7d6d]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(cling::Dyld::ContainsSymbol((anonymous namespace)::LibraryPath const*, llvm::StringRef, unsigned int) const+0x70) [0xb736e0]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(cling::Dyld::searchLibrariesForSymbol[abi:cxx11](llvm::StringRef, bool)+0x307) [0xb76cf7]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(cling::DynamicLibraryManager::searchLibrariesForSymbol[abi:cxx11](llvm::StringRef, bool) const+0x13) [0xb77053]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(AutoloadLibraryGenerator::tryToGenerate(llvm::orc::LookupState&, llvm::orc::LookupKind, llvm::orc::JITDylib&, llvm::orc::JITDylibLookupFlags, llvm::orc::SymbolLookupSet const&)+0x167) [0xb01be7]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::orc::ExecutionSession::OL_applyQueryPhase1(std::unique_ptr<llvm::orc::InProgressLookupState, std::default_delete<llvm::orc::InProgressLookupState> >, llvm::Error)+0x1f3) [0x2000503]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::orc::ExecutionSession::lookup(llvm::orc::LookupKind, std::vector<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags>, std::allocator<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags> > > const&, llvm::orc::SymbolLookupSet, llvm::orc::SymbolState, llvm::unique_function<void (llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::JITEvaluatedSymbol, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::JITEvaluatedSymbol> > >)>, std::function<void (llvm::DenseMap<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr> >, llvm::DenseMapInfo<llvm::orc::JITDylib*>, llvm::detail::DenseMapPair<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr> > > > const&)>)+0x34f) [0x200b1df]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::orc::ExecutionSession::lookup(std::vector<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags>, std::allocator<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags> > > const&, llvm::orc::SymbolLookupSet const&, llvm::orc::LookupKind, llvm::orc::SymbolState, std::function<void (llvm::DenseMap<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr> >, llvm::DenseMapInfo<llvm::orc::JITDylib*>, llvm::detail::DenseMapPair<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr> > > > const&)>)+0x19c) [0x200be0c]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::orc::ExecutionSession::lookup(std::vector<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags>, std::allocator<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags> > > const&, llvm::orc::SymbolStringPtr, llvm::orc::SymbolState)+0x112) [0x200c052]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::orc::LLJIT::lookupLinkerMangled(llvm::orc::JITDylib&, llvm::orc::SymbolStringPtr)+0xc2) [0x2022922]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(cling::IncrementalJIT::getSymbolAddress(llvm::StringRef, bool)+0x116) [0xc06666]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(cling::IncrementalExecutor::getAddressOfGlobal(llvm::StringRef, bool*) const+0x21) [0xbffd31]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(cling::Interpreter::getAddressOfGlobal(clang::GlobalDecl const&, bool*) const+0x6a) [0xb8571a]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(TClingDataMemberInfo::Offset()+0x157) [0xb1f367]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libcppyy_backend3_9.so(Cppyy::GetDatamemberOffset(unsigned long, unsigned long)+0x336) [0x13566]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libcppyy3_9.so(CPyCppyy::CPPDataMember::Set(unsigned long, unsigned long)+0x7b) [0x4eccb]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libcppyy3_9.so(CPyCppyy::meta_getattro(_object*, _object*)+0x1184) [0x5d9c4]
```

**Comment by vgvassilev** [comment]
> looks like root 628 and above are reading/opening all the shared libraries in LD_LIBRARY_PATH. e.g root version 6.28 and above load /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02808/el8_amd64_gcc12/external/blackhat/0.9.9-987ad1acae5cc088f5b0bffc0baf5368/lib/blackhat/libBG.so.0.0.0 which is from `blackhat` generator. In root 6.26 , on few of the root's shared libs were opened.

Yes, we index the symbols from system libraries generally. This feature was enabled in ROOT 626. I am not sure why 628 indexes more libraries than 626 as there have not been practically changes in that area since then.

I don't see how `ShallowAutoLoadImpl` could have changed since then either. Have you seen changes in the dictionaries in that area recently?

cc: @pcanal.

**Comment by Axel-Naumann** [comment]
@smuzaffar :
> all the shared libs from LD_LIBRARY_PATH and system directory (e.g. /usr/lib64) were opened (over 22K files were opened).

Are they opened *and not closed*?

**Comment by smuzaffar** [comment]
> Are they opened and not closed?

they were closed.

**Comment by Axel-Naumann** [comment]
Then I don't understand yet why the process is running out of file descriptors? Am I misunderstanding something?

**Comment by smuzaffar** [comment]
@Axel-Naumann , there were two issues

1. The cmssw unit test `testMakeTrackValidationPlots` was starting over 180 processes in parallel
2. ROOT 6.28 and above open over 22K files for each such process (though not simultaneously). Most of these files are read from cvmfs

Due to high number of parallel processes and large number of files opened by each process, the open file descriptors for `cvmfs` user hit the 1024 limit which resulted in errors like
```
Oct 21 16:07:35 cmsbuild114.cern.ch cvmfs2[6642]: (unpacked.cern.ch) open file descriptor limit exceeded
Oct 21 16:07:35 cmsbuild114.cern.ch cvmfs2[6642]: (unpacked.cern.ch) open file descriptor limit exceeded
```

(1) should be fixed by https://github.com/cms-sw/cmssw/pull/43096. But we still need to understand why ROOT 6.28 (and above) are opening all the shared libs in LD_LIBRARY_PATH +  system dirs.

**Comment by vgvassilev** [comment]
> But we still need to understand why ROOT 6.28 (and above) are opening all the shared libs in LD_LIBRARY_PATH + system dirs.

ROOT since at least 6.26 does pretty much what the system linker does but for all possible libraries on your LD_LIBRARY_PATH -- it searches for a symbol through all available libraries. The difference is that when the symbol is not found in the "usual" set of libraries then it goes off and looks at the system libraries which we observe here. I am not sure what changed across the two versions as this code hasn't. There might be several reasons I can think of:
  * The system libraries on that particular node became a lot more than they used to be;
  * We are searching for a symbol which existed in the past but does not exist now triggering the system search

**Comment by vgvassilev** [comment]
Ok... 626 uses LLVM13 and the orcv2 infrastructure tries more eagerly to respond on such symbol lookup requests.

@hahnjo is there a way to suppress the orcv2 callback in this stacktrace?
```
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(cling::DynamicLibraryManager::searchLibrariesForSymbol[abi:cxx11](llvm::StringRef, bool) const+0x13) [0xb77053]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(AutoloadLibraryGenerator::tryToGenerate(llvm::orc::LookupState&, llvm::orc::LookupKind, llvm::orc::JITDylib&, llvm::orc::JITDylibLookupFlags, llvm::orc::SymbolLookupSet const&)+0x167) [0xb01be7]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::orc::ExecutionSession::OL_applyQueryPhase1(std::unique_ptr<llvm::orc::InProgressLookupState, std::default_delete<llvm::orc::InProgressLookupState> >, llvm::Error)+0x1f3) [0x2000503]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::orc::ExecutionSession::lookup(llvm::orc::LookupKind, std::vector<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags>, std::allocator<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags> > > const&, llvm::orc::SymbolLookupSet, llvm::orc::SymbolState, llvm::unique_function<void (llvm::Expected<llvm::DenseMap<llvm::orc::SymbolStringPtr, llvm::JITEvaluatedSymbol, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr>, llvm::detail::DenseMapPair<llvm::orc::SymbolStringPtr, llvm::JITEvaluatedSymbol> > >)>, std::function<void (llvm::DenseMap<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr> >, llvm::DenseMapInfo<llvm::orc::JITDylib*>, llvm::detail::DenseMapPair<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr> > > > const&)>)+0x34f) [0x200b1df]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::orc::ExecutionSession::lookup(std::vector<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags>, std::allocator<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags> > > const&, llvm::orc::SymbolLookupSet const&, llvm::orc::LookupKind, llvm::orc::SymbolState, std::function<void (llvm::DenseMap<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr> >, llvm::DenseMapInfo<llvm::orc::JITDylib*>, llvm::detail::DenseMapPair<llvm::orc::JITDylib*, llvm::DenseSet<llvm::orc::SymbolStringPtr, llvm::DenseMapInfo<llvm::orc::SymbolStringPtr> > > > const&)>)+0x19c) [0x200be0c]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::orc::ExecutionSession::lookup(std::vector<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags>, std::allocator<std::pair<llvm::orc::JITDylib*, llvm::orc::JITDylibLookupFlags> > > const&, llvm::orc::SymbolStringPtr, llvm::orc::SymbolState)+0x112) [0x200c052]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(llvm::orc::LLJIT::lookupLinkerMangled(llvm::orc::JITDylib&, llvm::orc::SymbolStringPtr)+0xc2) [0x2022922]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(cling::IncrementalJIT::getSymbolAddress(llvm::StringRef, bool)+0x116) [0xc06666]
eb792bd7ce7d870/lib/libCling.so(cling::Interpreter::getAddressOfGlobal(clang::GlobalDecl const&, bool*) const+0x6a) [0xb8571a]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libCling.so(TClingDataMemberInfo::Offset()+0x157) [0xb1f367]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0eb792bd7ce7d870/lib/libcppyy_backend3_9.so(Cppyy::GetDatamemberOffset(unsigned long, unsigned long)+0x336) [0x13566]
 > <inst>/lcg/root/6.28.09-8486a0c5a951d64e0
 ```

My suspicion is that for some (static?) variable declaration we are requesting the offset which does not exist neither in the jit nor in the object files:

https://github.com/root-project/root/blob/a139a69cc27c186f8e1d765971a41f9f4dd43dc8/core/metacling/src/TClingDataMemberInfo.cxx#L355-L356

Perhaps that's related to what @wlav reported with recent cppyy and globals...

@smuzaffar I suspect we will need to print the mangled names. Can you apply this patch and rerun:
```diff
diff --git a/interpreter/cling/lib/Interpreter/Interpreter.cpp b/interpreter/cling/lib/Interpreter/Interpreter.cpp
index 681dbffc95..739f170d4e 100644
--- a/interpreter/cling/lib/Interpreter/Interpreter.cpp
+++ b/interpreter/cling/lib/Interpreter/Interpreter.cpp
@@ -1727,6 +1727,7 @@ namespace cling {
     // Return a symbol's address, and whether it was jitted.
     std::string mangledName;
     utils::Analyze::maybeMangleDeclName(GD, mangledName);
+    printf("mangled name=%s\n", mangledName.c_str());
 #if defined(_WIN32)
     // For some unknown reason, Clang 5.0 adds a special symbol ('\01') in front
     // of the mangled names on Windows, making them impossible to find
```
The output would be probably GBs...

**Comment by smuzaffar** [comment]
>     * The system libraries on that particular node became a lot more than they used to be;

@vgvassilev , tests were ran on same node so system libs should remain same. Note that root 6.26 is not reading any system or cmssw(and its dependencies) libs. ROOT 6.26 open just few root libs while root 6.28 opens over 22K shared libs.

>     * We are searching for a symbol which existed in the past but does not exist now triggering the system search

the symbol root tries to search must be from root itself and cmssw and all its externals are same. Only difference between CMSSW_13_3_X and CMSSW_13_3_ROOT628_X is root version

**Comment by smuzaffar** [comment]
@vgvassilev , I applied your patch and [a] are the only symbols which were searched

[a]
```
Singularity> grep 'mangled name' out.log  | sort | uniq -c
    501 mangled name=_ZL6kError
    501 mangled name=_ZL8kWarning
      2 mangled name=_ZN7TString5kNPOSE
      3 mangled name=gEnv
    501 mangled name=gErrorIgnoreLevel
      3 mangled name=gStyle
      3 mangled name=gSystem
```

**Comment by vgvassilev** [comment]
I guess there is no way to know which lookup triggers the round trip to the system libraries… Is that the same as what we got in 626?

**Comment by smuzaffar** [comment]
You can also run the following on lxplus9 to see the large number of libs loaded by root 6.28
```
> ssh lxplus9
> source /cvmfs/sft.cern.ch/lcg/views/dev4/latest/x86_64-el9-gcc12-opt/setup.sh
> strace -z -f -o run.log -e trace=open,openat python3 -c 'import ROOT;print(ROOT.gErrorIgnoreLevel)'
-1
> grep openat run.log | sed 's|.*, "||;s|".*||' | sort | uniq | wc -l
4415
```

**Comment by smuzaffar** [comment]
looks like any `ROOT.<name>` behaves the same way e.g `ROOT.kError` also opens all the shared libs in LD_LIBRARY_PATH.
Every unique `ROOT.<name>` go through all the libs again and again e.g. 
```
python3 -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);'
```
make 15K open calls and 
```
python3 -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.gErrorIgnoreLevel);'
```
make 23K open calls

**Comment by vgvassilev** [comment]
```nm $ROOTSYS/lib/*.so | grep _ZN7TString5kNPOSE
```

returns no results. The `kNPOS` is a static class variable: https://github.com/root-project/root/blob/1ed5e3947537dd3dcbd0eb5339a6c94e6afda23e/core/base/inc/TString.h#L280

That's another issue pointing to @wlav's reports.

The fact that we request the symbol `_ZN7TString5kNPOSE` which is `TString::kNPOS`. We cannot find it in the JIT land, we cannot find it in the ROOT libraries and that's why we are triggering an exhaustive search...

**Comment by smuzaffar** [comment]
@vgvassilev , do we already have any fix for [this](https://github.com/cms-sw/cmssw/issues/43077#issuecomment-1781310510)

**Comment by vgvassilev** [comment]
> @vgvassilev , do we already have any fix for [this](https://github.com/cms-sw/cmssw/issues/43077#issuecomment-1781310510)

@smuzaffar not yet. I was hoping that @hahnjo will have an idea why these symbols are invisible to the JIT.

**Comment by hahnjo** [comment]
> I was hoping that @hahnjo will have an idea why these symbols are invisible to the JIT.

No, I don't.

**Comment by wlav** [comment]
So two separate issues then on my side. One is problems with `std::string` which has a domino effect causing other problems. I've changed to a different workaround: explicit instantiation of `std::string` and `std::allocator<char>` in `clingwrapper.cxx`. That seems to be the easiest/most complete/portable solution.

The static symbols not being found seems to be unrelated in some cases and not necessarily a knock-on effect (although one report mentioned a load order problem: the library has to be loaded before the header). See also this one: https://github.com/wlav/cppyy/issues/156

**Comment by Axel-Naumann** [comment]
#14223 might be related.

**Comment by davidlange6** [comment]
Seems unlikely. Why so you suggest that?

**Comment by Axel-Naumann** [comment]
D'oh because I failed to link the issue correctly. Edited = fixed. Anyway, the rationale (for the *actual* candidate of "maybe related") is: https://github.com/cms-sw/cmssw/issues/43077#issuecomment-1777358996 shows how ROOT is opening 100 libs looking for a symbol. That takes time, and seems very similar to the diagnosis that Markus provided for the LHCb startup issue.

**Comment by makortel** [comment]
type root

**Comment by makortel** [comment]
In order for us to keep using ROOT 6.30, we'd need this issue fixed by the last open pre-release of CMSSW_14_0_0 (scheduled for Jan 23).

**Comment by smuzaffar** [comment]
https://github.com/root-project/root/pull/14261 seem to fix opening of all shared libs.

**Comment by smuzaffar** [comment]
I spoke too early, `python3 -c 'import ROOT;print(ROOT.kError);'` still opens all shared libs.

**Comment by smuzaffar** [comment]
thanks to root team specially @vepadulano , https://github.com/root-project/root/pull/14276 fixed the issue with python ROOT loading all the shared libs from LD_LIBRARY_PATH. This fix has been merged for root 6.28, 6.30 and master branches. [cmsdist](https://github.com/cms-sw/cmsdist) PRs

- ROOT628: https://github.com/cms-sw/cmsdist/pull/8903
- ROOT630 and CMSSW Production:  https://github.com/cms-sw/cmsdist/pull/8902
- ROOT6/master: https://github.com/cms-sw/cmsdist/pull/8900

have been merged for tonight's CMSSW_14_0_X IB.

**Comment by smuzaffar** [comment]
FYI, I ran all CMSSW unit tests using patch mentioned in https://github.com/cms-sw/cmssw/issues/43077#issuecomment-1779469244  and found out that unit tests are searching for symbols in [a,b,c]

- 3 ROOT symbols in [b] (with `*`) still trigger the searching of all shared libs. 
- All of CMSSW sysmbols triggers search all shared libs. I guess because they are declared `static` in header files.
  https://github.com/cms-sw/cmssw/blob/master/DataFormats/MuonReco/interface/Muon.h#L289-L296
  https://github.com/cms-sw/cmssw/blob/master/DataFormats/EgammaCandidates/interface/Photon.h#L565
  https://github.com/cms-sw/cmssw/blob/master/DataFormats/EgammaCandidates/interface/GsfElectron.h#L704
- One strange observation is that though [reco::GsfElectron::mvaPlaceholder](https://github.com/cms-sw/cmssw/blob/master/DataFormats/EgammaCandidates/interface/GsfElectron.h#L704) and [reco::Photon::mvaPlaceholder](https://github.com/cms-sw/cmssw/blob/master/DataFormats/EgammaCandidates/interface/Photon.h#L565) are set to `-999999999.` but `python3 -c 'import ROOT;print(ROOT.reco.GsfElectron.mvaPlaceholder)'` prints `-1000000000.0`. One can run `python3 ${CMSSW_BASE}/src/PhysicsTools/SelectorUtils/test/test_vid_selectors.py` for cmssw symbol search.

[a] cmssw
```
      2 reco::GsfElectron::mvaPlaceholder
      2 reco::Muon::CaloMuon
      2 reco::Muon::GEMMuon
      2 reco::Muon::GlobalMuon
      2 reco::Muon::ME0Muon
      2 reco::Muon::PFMuon
      2 reco::Muon::RPCMuon
      2 reco::Muon::StandAloneMuon
      2 reco::Muon::TrackerMuon
      2 reco::Photon::mvaPlaceholder
```
[b] ROOT (symbols with * still trigger searching all libs in LD_LIBRARY_PATH)
```
      4 TTreeReader::fgEntryStatusText
*     9 kFALSE
*    15 kTRUE
     21 gStyle
*    32 TTree::kMaxEntries
     66 TString::kNPOS
    111 gEnv
    111 gSystem
    501 kWarning
    507 gErrorIgnoreLevel
    507 kError
```
[c] std
```
     22 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::npos
```

**Comment by smuzaffar** [comment]
Note https://github.com/cms-sw/cmssw/issues/43077#issuecomment-1867640888 sis only for unit tests. I am running full relvals now but I do not expect we use python ROOT interface during Relvals

**Comment by vgvassilev** [comment]
If we use https://github.com/root-project/root/pull/14287 static constexpr variables should not need to go via the library symbol search.

**Comment by vgvassilev** [comment]
> * One strange observation is that though [reco::GsfElectron::mvaPlaceholder](https://github.com/cms-sw/cmssw/blob/master/DataFormats/EgammaCandidates/interface/GsfElectron.h#L704) and [reco::Photon::mvaPlaceholder](https://github.com/cms-sw/cmssw/blob/master/DataFormats/EgammaCandidates/interface/Photon.h#L565) are set to `-999999999.` but `python3 -c 'import ROOT;print(ROOT.reco.GsfElectron.mvaPlaceholder)'` prints `-1000000000.0`

Somebody dialed two too many digits for a float ;) https://godbolt.org/z/nYfo74rK1

**Comment by makortel** [comment]
I guess these

> * https://github.com/cms-sw/cmssw/blob/master/DataFormats/MuonReco/interface/Muon.h#L289-L296

should be changed to `static constexpr`.

**Comment by makortel** [comment]
@smuzaffar This issue has been resolved, right?

**Comment by smuzaffar** [comment]
@makortel , yes this has been fixed.

**Comment by makortel** [comment]
@cmsbuild, please close
